client/gc.go

```diff
@@ -0,0 +1,329 @@
+	}
+
+	gcAlloc := heap.Pop(&i.heap).(*GCAlloc)
+	delete(i.index, gcAlloc.allocRunner.Alloc().ID)
+	return gcAlloc
+}
+
+func (i *IndexedGCAllocPQ) Remove(allocID string) (*GCAlloc, error) {
+	i.pqLock.Lock()
+	defer i.pqLock.Unlock()
+
+	if gcAlloc, ok := i.index[allocID]; ok {
+		heap.Remove(&i.heap, gcAlloc.index)
+		delete(i.index, allocID)
+		return gcAlloc, nil
+	}
+
+	return nil, fmt.Errorf("alloc %q not present", allocID)
+}
+
+func (i *IndexedGCAllocPQ) Length() int {
+	i.pqLock.Lock()
+	defer i.pqLock.Unlock()
+
+	return len(i.heap)
+}
+
+// AllocGarbageCollector garbage collects terminated allocations on a node
+type AllocGarbageCollector struct {
+	allocRunners   *IndexedGCAllocPQ
+	statsCollector stats.NodeStatsCollector
+	reservedDiskMB int
+	logger         *log.Logger
+	shutdownCh     chan struct{}
+}
+
+// NewAllocGarbageCollector returns a garbage collector for terminated
+// allocations on a node.
+func NewAllocGarbageCollector(logger *log.Logger, statsCollector stats.NodeStatsCollector, reservedDiskMB int) *AllocGarbageCollector {
+	gc := &AllocGarbageCollector{
+		allocRunners:   NewIndexedGCAllocPQ(),
+		statsCollector: statsCollector,
+		reservedDiskMB: reservedDiskMB,
+		logger:         logger,
+		shutdownCh:     make(chan struct{}),
+	}
+	go gc.run()
+
+	return gc
+}
+
+func (a *AllocGarbageCollector) run() {
+	ticker := time.NewTicker(gcInterval)
+	for {
+		select {
+		case <-ticker.C:
+			if err := a.keepUsageBelowThreshold(); err != nil {
+				a.logger.Printf("[ERR] client: error GCing allocation: %v", err)
+			}
+		case <-a.shutdownCh:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// keepUsageBelowThreshold collects disk usage information and garbage collects
+// allocations to make disk space available.
+func (a *AllocGarbageCollector) keepUsageBelowThreshold() error {
+	for {
+		// Check if we have enough free space
+		err := a.statsCollector.Collect()
+		if err != nil {
+			return err
+		}
+
+		// See if we are below thresholds for used disk space and inode usage
+		diskStats := a.statsCollector.Stats().AllocDirStats
+
+		if diskStats == nil {
+			break
+		}
+
+		if diskStats.UsedPercent <= diskUsageThreshold &&
+			diskStats.InodesUsedPercent <= inodeUsageThreshold {
+			break
+		}
+
+		// Collect an allocation
+		gcAlloc := a.allocRunners.Pop()
+		if gcAlloc == nil {
+			break
+		}
+
+		ar := gcAlloc.allocRunner
+		alloc := ar.Alloc()
+		a.logger.Printf("[INFO] client: garbage collecting allocation %v", alloc.ID)
+
+		// Destroy the alloc runner and wait until it exits
+		ar.Destroy()
+		select {
+		case <-ar.WaitCh():
+		case <-a.shutdownCh:
+		}
+	}
+	return nil
+}
+
+func (a *AllocGarbageCollector) Stop() {
+	a.shutdownCh <- struct{}{}
+}
+
+// Collect garbage collects a single allocation on a node
+func (a *AllocGarbageCollector) Collect(allocID string) error {
+	gcAlloc, err := a.allocRunners.Remove(allocID)
+	if err != nil {
+		return fmt.Errorf("unable to collect allocation %q: %v", allocID, err)
```

Comment:
@dadgar The error line will have something like this as the reason - `alloc 123 not present`